### ISSUE TITLE
No error if you flash the wrong firmware

### ISSUE THREAD

**Issue by Adrienne200** [AllYarnsAreBeautiful/ayab-desktop#206]
Flash the 930 firmware, then use it on a 910, there should be an error. Or are we working towards a single firmware, or being OK with either 0.8 firmware?

On a 910, flashed with 930etc 0.8 firmware, 0.9 software:
Set up to knit a simple pattern, default settings. 
Expected: It should error when you press Knit.
What’s actually happening is that it half-works; it selects row 1 correctly and beeps, fails to select row 2 and no beep even if I cross the turn mark, then on the third pass it selects row 2 and beeps. Basically mostly only working on the left-right passes. Only sometimes beeps on the left.
Tried a current EMSL board and a Tinkersoup original, same.
Log indicates that the data is being sent down correctly.
Looks like it did this in 0.8 too on this machine and I hadn’t noticed. 0.75 does give the error that I expected; if it showed now it should have a wording change, #127 

The reverse case, I'm not sure. 910 firmware on a 930 machine also doesn't give the error, but it appears to work as usual, at least in a quick basic test. 
Then one more test, 930 fw on a different 910 machine. First test worked, second time through it failed as above.

[ayab_log 830 fw on 910.txt](https://github.com/AllYarnsAreBeautiful/ayab-desktop/files/1100548/ayab_log.830.fw.on.910.txt)

**Comment by oskay** [comment]
It is probably a good idea to query the firmware version and configuration. I'd recommend that we target this as a milestone post 0.9.

**Comment by chris007de** [comment]
@oskay we are already querying the firmware version at the very beginning of the serial communication. If the firmware version is not as expected, the GUI shows an error.
Also, I would probably remove any non-compatible firmware images from the release, so people cannot even flash images which are incompatible with the GUI they are currently running.

@Adrienne200 
We were already working on some kind of "monolithic" firmware which covers all machine models (https://github.com/allyarnsarebeautiful/ayab-desktop/tree/v0.9_mono_firmware), but I put it on hold as we started doing serious changes (python3, pyqt5).
The goal was to remove the need to decide for a machine model when flashing the Arduino, so we only have a single firmware hex file to distribute (of course, there would be still separate images for different Arduino models such as Uno, Mega).
But, as all the machine models 910, 930, ... etc are electrically the same, there is currently no chance for the Arduino to recognize the machine model it is connected to by its own (at least I am not aware of any possibility). But the Arduino has to act different for the different models, as the hall sensors give different values on different machines (https://github.com/AllYarnsAreBeautiful/ayab-firmware/blob/master/settings.h#L41). So even with such a "one-firmware-fits-all"-firmware, we end up with the need to choose the machine model manually before we start knitting. This still yields potential for misconfiguration, and one would not notice until knitting fails with the behavior you describe.
Perhaps there would be possibilities to do a machine model detection, e.g. adding a "recognition" phase during before knitting where user has to pass the carriage past both hall sensors first, so the firmware can choose the right parameters, or we do hardware changes where we add a DIP switch to hard-code the machine type or think of a way to recognize(/guess) the connected machine type by sensing which connector ports are used. But I guess we have to weigh measures against each other...how much technical effort do we want to spend to help even the most inattentive user?

tl;dr:
The plan is to chose the machine model no longer at the time of firmware flashing, but before every time you start to knit, perhaps that solves the problem already, as users are more aware of their choice.

**Comment by dl1com** [comment]
> This will become moot with the next, unitary firmware release.

Fixed by 1.0
